File: examples/Actor.Next/05-Interpreted/DeviceManagementDemo.cs
// ------------------------------------------------------------------------
// Copyright 2026 The Dapr Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

using Dapr.Actors.Next.Abstractions.Filters;
using Dapr.Actors.Next.Abstractions.Registry;
using Dapr.Actors.Next.Interpreted;
using System.Text.Json;

namespace Dapr.Actors.Next.Examples.DeviceManagement;

public static class DeviceManagementDemo
{
    public const string SmartLockDeviceType = "SmartLock";

    public static InterpretedMachineDefinition SmartLockDefinition(bool includeUnlockingCompletion = true, string motorEffect = "ActuateMotor") =>
        new()
        {
            DocumentVersion = 1,
            InitialState = "Locked",
            States =
            [
                new InterpretedStateDefinition { Name = "Locked" },
                new InterpretedStateDefinition { Name = "Unlocking" },
                new InterpretedStateDefinition { Name = "Unlocked" },
                new InterpretedStateDefinition { Name = "Locking" },
            ],
            Transitions = Transitions(includeUnlockingCompletion, motorEffect),
        };

    public static InterpretedMachineVerificationResult Verify(
        IInterpretedMachineVerifier verifier,
        InterpretedMachineDefinition definition) =>
        verifier.Verify(definition);

    private static IReadOnlyList<InterpretedTransitionDefinition> Transitions(bool includeUnlockingCompletion, string motorEffect)
    {
        var transitions = new List<InterpretedTransitionDefinition>
        {
            new()
            {
                Source = "Locked",
                Event = "Unlock",
                Branches =
                [
                    new InterpretedBranchDefinition
                    {
                        Guards = ["CheckBattery"],
                        Target = "Unlocking",
                        Effects = [motorEffect],
                    },
                ],
            },
            new()
            {
                Source = "Unlocked",
                Event = "Lock",
                Branches =
                [
                    new InterpretedBranchDefinition
                    {
                        Otherwise = true,
                        Target = "Locking",
                        Effects = [motorEffect],
                    },
                ],
            },
            new()
            {
                Source = "Locking",
                Event = "MotorStopped",
                Branches =
                [
                    new InterpretedBranchDefinition
                    {
                        Otherwise = true,
                        Target = "Locked",
                    },
                ],
            },
        };

        if (includeUnlockingCompletion)
        {
            transitions.Add(new InterpretedTransitionDefinition
            {
                Source = "Unlocking",
                Event = "MotorStopped",
                Branches =
                [
                    new InterpretedBranchDefinition
                    {
                        Otherwise = true,
                        Target = "Unlocked",
                    },
                ],
            });
        }

        return transitions;
    }
}

public sealed class DeviceControlPlane(IActorRegistry registry, IDynamicActorClient client)
{
    public IReadOnlyList<string> ListDeviceOperations() =>
        registry.Actors
            .SelectMany(actor => actor.Methods.Select(method => $"{actor.ActorType}.{method.Name}({string.Join(", ", method.Parameters.Select(parameter => parameter.ParameterType.Name))})"))
            .Order(StringComparer.Ordinal)
            .ToArray();

    public Task<string?> SendCommandAsync(string deviceType, string deviceId, string command, CancellationToken cancellationToken = default) =>
        RaiseAsync(deviceType, deviceId, command, new DeviceCommand(command), cancellationToken);

    public Task<string?> RecordSensorEventAsync(string deviceType, string deviceId, string sensorEvent, CancellationToken cancellationToken = default) =>
        RaiseAsync(deviceType, deviceId, sensorEvent, new DeviceSensorEvent(sensorEvent), cancellationToken);

    private Task<string?> RaiseAsync<TPayload>(
        string deviceType,
        string deviceId,
        string eventName,
        TPayload payload,
        CancellationToken cancellationToken)
    {
        var evt = new InterpretedEvent(eventName, JsonSerializer.SerializeToElement(payload));
        return client.InvokeAsync(deviceType, deviceId, "Raise", JsonSerializer.Serialize(evt), cancellationToken);
    }
}

public sealed class DeviceTypeRegistry : IActorRegistry
{
    private readonly Dictionary<string, ActorTypeDescriptor> devices = new(StringComparer.Ordinal);

    public IReadOnlyList<ActorTypeDescriptor> Actors => devices.Values.OrderBy(device => device.ActorType, StringComparer.Ordinal).ToArray();

    public void AddDeviceType(string deviceType)
    {
        devices[deviceType] = new ActorTypeDescriptor(
            deviceType,
            1,
            typeof(InterpretedStateMachineActor),
            typeof(InterpretedStateMachineActor),
            [
                new ActorMethodDescriptor(
                    "Raise",
                    "Raise",
                    typeof(InterpretedRaiseResult),
                    [new ActorParameterDescriptor("evt", typeof(InterpretedEvent), 0, false, false, null)]),
            ]);
    }

    public bool TryGet(string actorType, out ActorTypeDescriptor descriptor) => devices.TryGetValue(actorType, out descriptor!);
}

public sealed class DeviceCapabilityRegistry : ICapabilityRegistry
{
    private readonly Dictionary<string, IActorEffect> effects = new(StringComparer.Ordinal);
    private readonly Dictionary<string, IActorGuard> guards = new(StringComparer.Ordinal);

    public DeviceCapabilityRegistry()
    {
        effects["ActuateMotor"] = new MotorEffect();
        guards["CheckBattery"] = new BatteryGuard();
    }

    public bool TryGetEffect(string name, out IActorEffect effect) => effects.TryGetValue(name, out effect!);

    public bool TryGetGuard(string name, out IActorGuard guard) => guards.TryGetValue(name, out guard!);

    private sealed class MotorEffect : IActorEffect
    {
        public ValueTask ExecuteAsync(ActorCapabilityContext context, CancellationToken cancellationToken = default) =>
            ValueTask.CompletedTask;
    }

    private sealed class BatteryGuard : IActorGuard
    {
        public ValueTask<bool> EvaluateAsync(ActorCapabilityContext context, CancellationToken cancellationToken = default) =>
            ValueTask.FromResult(true);
    }
}

public sealed record DeviceCommand(string Name);

public sealed record DeviceSensorEvent(string Name);

File: examples/Actor.Next.FSharp/05-Interpreted/Device.Next.FSharp.Example05/DeviceManagementDemo.fs
#nowarn "FS3261"
namespace Dapr.Actors.Next.Examples.DeviceManagement

open System
open System.Collections.Generic
open System.Linq
open System.Runtime.InteropServices
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open Dapr.Actors.Next.Abstractions
open Dapr.Actors.Next.Abstractions.Filters
open Dapr.Actors.Next.Abstractions.Registry
open Dapr.Actors.Next.Interpreted

type DeviceCommand = { Name: string }
type DeviceSensorEvent = { Name: string }
type DeviceTypeDescription = { ActorType: string; Operations: IReadOnlyList<string> }
type DeviceOnboarded = { DeviceType: string; DeviceId: string; InitialState: string }

module DeviceManagementDemo =
    let SmartLockDeviceType = "SmartLock"

    let SmartLockDefinition (includeUnlockingCompletion: bool) (motorEffect: string) : InterpretedMachineDefinition =
        let transitions = ResizeArray<InterpretedTransitionDefinition>()

        transitions.Add(InterpretedTransitionDefinition(
            Source = "Locked",
            Event = "Unlock",
            Branches = [|
                InterpretedBranchDefinition(
                    Guards = [ "CheckBattery" ],
                    Target = "Unlocking",
                    Effects = [ motorEffect ]
                )
            |]
        ))

        transitions.Add(InterpretedTransitionDefinition(
            Source = "Unlocked",
            Event = "Lock",
            Branches = [|
                InterpretedBranchDefinition(
                    Otherwise = true,
                    Target = "Locking",
                    Effects = [ motorEffect ]
                )
            |]
        ))

        transitions.Add(InterpretedTransitionDefinition(
            Source = "Locking",
            Event = "MotorStopped",
            Branches = [|
                InterpretedBranchDefinition(
                    Otherwise = true,
                    Target = "Locked"
                )
            |]
        ))

        if includeUnlockingCompletion then
            transitions.Add(InterpretedTransitionDefinition(
                Source = "Unlocking",
                Event = "MotorStopped",
                Branches = [|
                    InterpretedBranchDefinition(
                        Otherwise = true,
                        Target = "Unlocked"
                    )
                |]
            ))

        InterpretedMachineDefinition(
            DocumentVersion = 1,
            InitialState = "Locked",
            States = [|
                InterpretedStateDefinition(Name = "Locked")
                InterpretedStateDefinition(Name = "Unlocking")
                InterpretedStateDefinition(Name = "Unlocked")
                InterpretedStateDefinition(Name = "Locking")
            |],
            Transitions = transitions.ToArray()
        )

    let SmartLockDefault () = SmartLockDefinition true "ActuateMotor"

    let Verify (verifier: IInterpretedMachineVerifier) (definition: InterpretedMachineDefinition) : InterpretedMachineVerificationResult =
        verifier.Verify(definition)

type DeviceControlPlane(registry: IActorRegistry, client: IDynamicActorClient) =

    member _.ListDeviceOperations() : IReadOnlyList<string> =
        registry.Actors
            .SelectMany(fun actor ->
                actor.Methods.Select(fun method ->
                    let paramList = String.Join(", ", method.Parameters.Select(fun p -> p.ParameterType.Name))
                    actor.ActorType + "." + method.Name + "(" + paramList + ")"))
            .Order(StringComparer.Ordinal)
            .ToArray()

    member _.SendCommandAsync(deviceType: string, deviceId: string, command: string, ?cancellationToken: CancellationToken) : Task<string> =
        let ct = defaultArg cancellationToken CancellationToken.None
        let payload = { DeviceCommand.Name = command }
        let evt = InterpretedEvent(command, JsonSerializer.SerializeToElement(payload))
        client.InvokeAsync(deviceType, deviceId, "Raise", JsonSerializer.Serialize(evt), ct)

    member _.RecordSensorEventAsync(deviceType: string, deviceId: string, sensorEvent: string, ?cancellationToken: CancellationToken) : Task<string> =
        let ct = defaultArg cancellationToken CancellationToken.None
        let payload = { DeviceSensorEvent.Name = sensorEvent }
        let evt = InterpretedEvent(sensorEvent, JsonSerializer.SerializeToElement(payload))
        client.InvokeAsync(deviceType, deviceId, "Raise", JsonSerializer.Serialize(evt), ct)

type DeviceTypeRegistry() =
    let devices = Dictionary<string, ActorTypeDescriptor>(StringComparer.Ordinal)

    member _.AddDeviceType(deviceType: string) =
        devices.[deviceType] <- ActorTypeDescriptor(
            deviceType, 1,
            typeof<InterpretedStateMachineActor>,
            typeof<InterpretedStateMachineActor>,
            [
                ActorMethodDescriptor(
                    "Raise", "Raise",
                    typeof<InterpretedRaiseResult>,
                    [ ActorParameterDescriptor("evt", typeof<InterpretedEvent>, 0, false, false, null) ]
                )
            ])

    interface IActorRegistry with
        member _.Actors = devices.Values.OrderBy(fun d -> d.ActorType, StringComparer.Ordinal).ToArray() :> IReadOnlyList<ActorTypeDescriptor>
        member _.TryGet(actorType: string, [<Out>] descriptor: byref<ActorTypeDescriptor>) : bool =
            let ok, v = devices.TryGetValue(actorType)
            if ok then
                descriptor <- v
                true
            else
                false

type DeviceCapabilityRegistry() =
    let effects = Dictionary<string, IActorEffect>(StringComparer.Ordinal)
    let guards = Dictionary<string, IActorGuard>(StringComparer.Ordinal)

    do
        effects.["ActuateMotor"] <- { new IActorEffect with
            member _.ExecuteAsync(_: ActorCapabilityContext, _: CancellationToken) : ValueTask = ValueTask.CompletedTask }
        guards.["CheckBattery"] <- { new IActorGuard with
            member _.EvaluateAsync(_: ActorCapabilityContext, _: CancellationToken) : ValueTask<bool> = ValueTask.FromResult(true) }

    interface ICapabilityRegistry with
        member _.TryGetEffect(name: string, [<Out>] effect: byref<IActorEffect>) : bool =
            let ok, v = effects.TryGetValue(name)
            if ok then effect <- v; true else false
        member _.TryGetGuard(name: string, [<Out>] guard: byref<IActorGuard>) : bool =
            let ok, v = guards.TryGetValue(name)
            if ok then guard <- v; true else false
